### PR TITLE
Supplement spring-retry listener features

### PR DESCRIPTION
**Pull request by holy12345**
Hi @ryanjbaxter 
Thank you very much for your patience guidance, and I have re-submitted PR as you suggested (PR #293 has been closed before).
Allow me to explain this PR content

- Add spring-retry listener features

- Added test code for this feature

I hope you can help me check, any of your reply is a great encouragement to me. Thanks again : )

### CODE REVIEW

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/RetryLoadBalancerInterceptor.java [hunk 1]
```diff
@@ -90,6 +96,10 @@ public ClientHttpResponse intercept(final HttpRequest request, final byte[] body
 		BackOffPolicy backOffPolicy = backOffPolicyFactory.createBackOffPolicy(serviceName);
 		template.setBackOffPolicy(backOffPolicy == null ? new NoBackOffPolicy() : backOffPolicy);
 		template.setThrowLastExceptionOnExhausted(true);
+        RetryListener[] retryListeners = this.retryListenerFactory.createRetryListeners(serviceName);
```
Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/RetryLoadBalancerInterceptor.java]: Spacing looks like it is wrong here

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/RetryLoadBalancerInterceptor.java [hunk 2]
```diff
@@ -64,18 +67,21 @@ public RetryLoadBalancerInterceptor(LoadBalancerClient loadBalancer,
...
+		this.retryListenerFactory = new LoadBalancedRetryListenerFactory.NoRetryListenerFactory();
 	}
 
 	public RetryLoadBalancerInterceptor(LoadBalancerClient loadBalancer,
 										LoadBalancerRetryProperties lbProperties,
 										LoadBalancedRetryPolicyFactory lbRetryPolicyFactory,
 										LoadBalancerRequestFactory requestFactory,
-										LoadBalancedBackOffPolicyFactory backOffPolicyFactory) {
+										LoadBalancedBackOffPolicyFactory backOffPolicyFactory,
+                                        LoadBalancedRetryListenerFactory retryListenerFactory) {
```
Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/RetryLoadBalancerInterceptor.java]: Add a new constructor so we dont break anyone.  Mark this one as deprecated and should be removed in 2.0.x

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalancedRetryListenerFactory.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Factory class to return the retry listeners.
+ * @author Gang Li
+ */
+public interface LoadBalancedRetryListenerFactory {
+
+    RetryListener[] createRetryListeners(String service);
+
+    class NoRetryListenerFactory implements LoadBalancedRetryListenerFactory {
```
Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalancedRetryListenerFactory.java]: I would rather this be called `DefaultRetryListenerFactory`

**Comment by holy12345** [comment]
Hi @ryanjbaxter 
It's fantastic for me to get your reply, and I'll modify it now. thank you very much

**Comment by holy12345** [comment]
In accordance with your suggestion I have completed the revision, please check again. Thanks again : )

**Pull request merged**